new quickstart

ibdthor opened the pull request:
new quickstart

ibdthor committed:
new quickstart

new quickstart

ibdknox commented on v0.3/tutorials/quickstart.eve:
````diff
@@ -7,117 +7,95 @@ title: Quickstart
-Searches find records that matches the supplied pattern, in this case records tagged greeting with a message attribute. You use these results to create new records. In this block, we create a `#ui/text` record for every message that is found (the UI library looks for records tagged `#ui/text` and draws them as DOM elements in the browser). If no records match the search, the block has no effect.  ...
+```
+commit
+  [#ui/div #logo class:"qs-logo" sort: 1]
````
if you wrap this in an outer div, you won't need the explicit sorts

ibdthor replied:
had one originally, traded sorts for nesting. should i put it back?

joshuafcole commented on v0.3/tutorials/quickstart.eve:
````diff
@@ -7,117 +7,95 @@ title: Quickstart
 
-Let's start with a block that adds a record to Eve (we'll show the code fence in this block, but elide them in later blocks):
+```
+npm start
````
need to add escaped ticks

joshuafcole commented on v0.3/tutorials/quickstart.eve:
````diff
@@ -7,117 +7,95 @@ title: Quickstart
-## Finding records in Eve
 
-Blocks search for records, and then bind or commit new records. Let's search for the `#greeting` we just committed, and then display it on the screen:
+```
````
need to add escaped ticks

joshuafcole commented on v0.3/tutorials/quickstart.eve:
````diff
@@ -7,117 +7,95 @@ title: Quickstart
+```
 
-Variables with the same name are equivalent within a block; because they have the same name, the `message` in `[#greeting message]` and `[#ui/text text: message]` are equivalent.
+The attribute `class` gets added to each div, which already has a little premade CSS to help see what we're doing. The logo is present, but next we need to add some pages to our nav bar and give each page some content.
````
Instead of passive voice "gets added to", I'd make it active. "We add a class attribute to each div... to help see what we're doing"

ibdthor replied:
used passive because the user's not doing it, eve is, but i changed it to active and explicitly said what eve is doing

joshuafcole commented on v0.3/tutorials/quickstart.eve:
````diff
@@ -7,117 +7,95 @@ title: Quickstart
+Before you start with this tutorial, please follow the [installation and usage instructions](/v0.3/install/), which will get you running Eve programs on your machine. In the eve-starter directory, you can launch the Program Switcher:
 
-Let's start with a block that adds a record to Eve (we'll show the code fence in this block, but elide them in later blocks):
+```
````
this code fence should be 
````
``` sh
````

joshuafcole replied:
This is still an issue. Otherwise the docs highlighter will think this block is meant to be eve code.

joshuafcole commented on v0.3/tutorials/quickstart.eve:
````diff
@@ -7,117 +7,95 @@ title: Quickstart
+```
 
-As the time changes, the output updates to reflect the current state of the `#time` record. Records can be committed instead of bound, but the behavior is a little different -- committed records persist until they are removed explicitly.
+These records alone don't add anything to the browser though. To actually make them appear, we need to write a block that first looks for our pages, then adds them into the browser.
````
Mentioning this at the end has the same issue of users thinking they did something wrong and trying to debug before they continue reading. Perhaps we should move this note above the block?

joshuafcole commented on v0.3/tutorials/quickstart.eve:
```diff
@@ -7,117 +7,95 @@ title: Quickstart
-## Summary
-
-That's it for the 5 minute introduction to Eve. To summarize:
+  view.children += ([#ui/div text: "This is the Eve Quickstart app! Choose a link to see some of the things Eve can do."], [#ui/img src:"http://i.imgur.com/qup3Y7l.jpg"])
```
I'd put those records on newlines. This looks pretty cluttered.

joshuafcole commented on v0.3/tutorials/quickstart.eve:
```diff
@@ -7,117 +7,95 @@ title: Quickstart
-- Bound records are replaced when data changes, while committed records must be removed manually.
-- Records are unique, uniqueness is determined by a record's attributes and their values.
+## Navigation
+The last thing to do is make the nav bar links switch which page the app is showing.
```
We need to address the fact that events are cleaned up here, and that they're getting sourced from a built in library.

ibdthor replied:
strongly disagree. trying to teach eve concepts in the qs is what got us into trouble to begin with - here, they expect to click a button and see the page change, which is what happens, so we shouldn't clutter it up with explanations for things that already make sense. click event clean-ups get taught very shortly down the road with the counter example, which fits in more organically

joshuafcole replied:
Where is the flow to that? Right now I don't see one.

joshuafcole reviewed (changes requested):
few preliminary notes

ibdthor committed:
cleaned up qs and added next steps

next steps: clock, counter, improved counter, 1 button stopwatch, and 2
button stopwatch

cmontella commented on v0.3/tutorials/quickstart 1 button stopwatch.eve:
````diff
@@ -0,0 +1,198 @@
+  [#race-timer #stopped #reset]
+```
+
+Stopwatches tend to measure seconds to two decimal points, so our resolution changes to 10ms (or one centisecond) to account for that. The other thing added here is a `#race-timer` record that lets us keep track of when the stopwatch is stopped or running or has been reset back to zero.
````
"stopped or running or has been reset back to zero." > "stopped, running, or reset."

joshuafcole replied:
:+1:

cmontella commented on v0.3/tutorials/quickstart 1 button stopwatch.eve:
```diff
@@ -0,0 +1,198 @@
+
+### Draw the button
+
+All we need to do is add one button to cycle between start, stop, and reset.
```
"All we need to do is add.." > "We add..."

cmontella commented on v0.3/tutorials/quickstart 1 button stopwatch.eve:
````diff
@@ -0,0 +1,198 @@
+  race-timer += #running
+```
+
+If the button is pressed while the stopwatch is stopped and reset (thus at zero), this records the `#start-time` and changes the `#race-timer` from `#stopped` to `#running` and removes the `#reset` tag.
````
Our code-style so far has been to write prose before the code block it describes.

ibdthor replied:
and so far that has worked out poorly. people want to copy/paste, then figure it out

cmontella commented on v0.3/tutorials/quickstart 1 button stopwatch.eve:
````diff
@@ -0,0 +1,198 @@
+  race-timer.millisecond += 0
+```
+
+Before the stopwatch button is pressed, there are no time values anywhere, so we just manually set three attributes to the `#race-timer` for minutes, seconds, and milliseconds, and make those all 0. This way, when the page is loaded or the stopwatch is reset, the hands are zeroed out.
````
"The stopwatch has no values before the start button is pressed, so we add default values for the minute, second, and millisecond attributes."

joshuafcole replied:
My previous comments here rescinded, your overall method is something I had not expected that works fine.

cmontella commented on v0.3/tutorials/quickstart 2 button stopwatch.eve:
`````diff
@@ -0,0 +1,252 @@
+## Implement system time
+
+````
+```
`````
we should probably update the code blocks to v0.4 style. This way, we don't have to talk about embedding them and showing nested code fences like this.

joshuafcole replied:
You monsters

cmontella commented on v0.3/tutorials/quickstart 2 button stopwatch.eve:
```diff
@@ -0,0 +1,252 @@
+
+# Quickstart + Two Button Stopwatch
+
+This guide expands on the quickstart `Clock` page and shows you how to turn the clock into a two button stopwatch. The following code is meant to replace all the code from the original clock.
```
Link to the original

cmontella commented on v0.3/tutorials/quickstart clock.eve:
```diff
@@ -0,0 +1,145 @@
+
+This is made to add onto the web app you made in the quickstart. You'll want to add the new code from here into the quickstart.eve document if you've already got the web app written there. Otherwise, you can quickly copy and paste the code in this section to recreate it. Whichever path you take, jump down to the `Building the clock` section to begin working on the counter!
+
+Header
```
This one could use more instruction. It's not clear to me what the story is past this point.

joshuafcole replied:
Maybe instead of re-embedding it here, link them back to that document?

cmontella commented on v0.3/tutorials/quickstart counter solution.eve:
`````diff
@@ -0,0 +1,42 @@
+```
+````
+
+A common trope is that good programmers delete code, and this follows suit - we actually get a more elegant solution here by condensing two blocks before down into one. We can't just smash together the two blocks entirely though because we want the original message to read "The button has been clicked 0 times" before any clicks happen. Recall that in the original counter example, before any click ...
`````
"A common trope is that good programmers delete code, and this follows suit" we don't have to make a value judgement here

joshuafcole replied:
That doesn't seem like a big deal to me either way

joshuafcole commented on v0.3/tutorials/quickstart 1 button stopwatch.eve:
````diff
@@ -0,0 +1,198 @@
+    [#second-hand sort: 4, degrees: 6 * second + 0.006 * millisecond, length: 46, stroke: "#ff0000", stroke-width: 0.5]]
+```
+
+This block closely resembles its predecessor, but the search looks for `#race-timer` now instead of the system time since it's the race time that will determine the position of the hands.
````
I would explain that the timer is a record like any other, and you're creating a timer for the stopwatch that ticks much faster than the previous timer which you've tagged #race-timer.

ibdthor replied:
it's unclear to me what that description adds here. the system timer is ticking at the same speed as the race timer; the race timer is just counting from 0 as opposed to jan 1 1970

joshuafcole commented on v0.3/tutorials/quickstart 1 button stopwatch.eve:
````diff
@@ -0,0 +1,198 @@
+  race-timer.millisecond += 0
+```
+
+Before the stopwatch button is pressed, there are no time values anywhere, so we just manually set three attributes to the `#race-timer` for minutes, seconds, and milliseconds, and make those all 0. This way, when the page is loaded or the stopwatch is reset, the hands are zeroed out.
````
This code isn't actually setting a default, you should end up with two values whenever you reset, unless you're handling the removal of the other values elsewhere, That may incidentally work since the SVG guy will overwrite attributes last write wins, but it's probably not a style we should encourage.

joshuafcole replied:
I suppose this is fine since you never actually add #reset, but in that case I'd probably choose a clearer tag name

joshuafcole commented on v0.3/tutorials/quickstart 2 button stopwatch.eve:
```diff
@@ -0,0 +1,252 @@
+  race := none
+  start := none
+  stop := none
+  [#race-timer #stopped #reset]
```
This strategy works, scratch my comments about it not from earlier :+1:

joshuafcole commented on v0.3/tutorials/quickstart 2 button stopwatch.eve:
`````diff
@@ -0,0 +1,252 @@
+### Subsequent stops
+
+````
+```
`````
Any reason not to just nuke and recreate it?

ibdthor replied:
yes

joshuafcole replied:
While yes is technically an answer I was hoping you might expound for my benefit

joshuafcole commented on v0.3/tutorials/quickstart clock.eve:
`````diff
@@ -0,0 +1,145 @@
+```
+````
+
+## Building the clock
`````
Seems like this should be H2 (as it is) but the blocks beneath should be H3. They're all part of building the clock, right?

joshuafcole commented on v0.3/tutorials/quickstart counter.eve:
```diff
@@ -0,0 +1,148 @@
+
+This guide expands on the quickstart by adding a simple incrementing counter to the `Counter` page.
+
+## Quickstart code
```
I would also link out rather than re-embed. The examples should be able to stand on their own or be embedded into the quickstart.

ibdthor committed:
updates, links, and new file names

ibdthor committed:
corrected links

ibdthor committed:
changed #race-timer to #stopwatch

joshuafcole committed:
Fix merge conflict

joshuafcole committed:
Refix merge
